lib/Parse/Lexer.cpp

```diff
@@ -1477,6 +1477,29 @@ void Lexer::lexImpl() {
       } else {
         diagnose(CurPtr-1, diag::lex_invalid_character)
           .fixItReplaceChars(getSourceLoc(CurPtr-1), getSourceLoc(tmp), " ");
+
+        StringRef confusedChar;
+        if (codepoint >= 0x80) {
+          llvm::SmallString<64> buffer;
+          EncodeToUTF8(codepoint, buffer);
+          confusedChar = buffer.str();
+        } else {
+          confusedChar = StringRef(std::string(1, (char)codepoint));
```

Comment:
StringRef is a non-owning type, so both branches here result in references to deallocated stack memory. We generally think it's okay to make the diagnostic path slower, so just taking the slow path all the time here is okay rather than trying to be clever.